### PR TITLE
Upgrade to ruby 3.0

### PR DESCRIPTION
**Pull request by zendesk-mattlefevre**
## Description
The time has come: We're working on bumping Samson to Ruby 3.0

### Risks
- High - 2.7 to 3.0 is a big jump with some breaking changes.

### CODE REVIEW

**Comment by grosser** [comment]
best do the gem upgrades in 1-off PRs to lower the risk / complexity🤞

File: .ruby-version [hunk 1]
```diff
@@ -1 +1 @@
-2.7.6
+3.0.6
```
Comment (grosser) [inline .ruby-version:1]: make sure that is available on the samson hosts ... last 2.7.8 upgrade failed on that

File: app/controllers/access_tokens_controller.rb [hunk 2]
```diff
@@ -48,7 +48,8 @@ def resource_class
 
   def render_resource_as_json(**args)
     # Override constraints in Doorkeeper::AccessToken
-    render_as_json resource_name, @resource.serializable_hash, nil, **args, allowed_includes: nil
+    kwargs = args.merge({allowed_includes: nil})
+    render_as_json resource_name, @resource.serializable_hash, nil, **kwargs
```
Comment (grosser) [inline app/controllers/access_tokens_controller.rb:52]:
```suggestion
    render_as_json resource_name, @resource.serializable_hash, nil, allowed_includes: nil, **args
```

File: app/controllers/concerns/json_renderer.rb [hunk 3]
```diff
@@ -57,7 +57,7 @@ def render_as_json(namespace, resource, pagy, status: :ok, allowed_includes: [])
       end
 
     # prepare resources for include collection
-    resource_json = resource.as_json(methods: requested_inlines)
+    resource_json = resource.as_json(**{methods: requested_inlines})
```
Comment (grosser) [inline app/controllers/concerns/json_renderer.rb:60]: that seems like it's exactly the same as before with more syntax ...

File: app/models/release_service.rb [hunk 4]
```diff
@@ -23,7 +23,7 @@ def can_release?
   private
 
   def push_tag_to_git_repository(version, commit)
-    GITHUB.create_release(@project.repository_path, version, target_commitish: commit)
+    GITHUB.create_release(@project.repository_path, version, **{target_commitish: commit})
```
Comment (grosser) [inline app/models/release_service.rb:26]:
```suggestion
    GITHUB.create_release(@project.repository_path, version, target_commitish: commit)
```

File: config/initializers/doorkeeper.rb [hunk 5]
```diff
@@ -14,9 +14,9 @@ module Doorkeeper
   module Orm
     module ActiveRecord
       class << self
-        undef initialize_models!
-        def initialize_models!
-        end
+        # undef initialize_models!
```
Comment (grosser) [inline config/initializers/doorkeeper.rb:17]: remove code or restore ?

File: config/initializers/faraday.rb [hunk 6]
```diff
@@ -7,6 +7,7 @@
   def run_request(method, url, *args)
     raise "Missing url for logging, do not use `do |request| request.url = ` pattern" unless url
     log_url = url.gsub(/key=\w+/, "key=redacted") # ignore datadog credentials and maybe others
-    ActiveSupport::Notifications.instrument("request.faraday.samson", method: method, url: log_url) { super }
+    payload = {method: method, url: log_url}
```
Comment (grosser) [inline config/initializers/faraday.rb:10]: don't understand why this was extracted ... line too long ?

File: lib/samson/secrets/vault_server.rb [hunk 7]
```diff
@@ -72,15 +72,14 @@ def sync!(other)
       end
 
       def create_client
-        VaultClientWrapper.new(
+        VaultClientWrapper.new(**
           DEFAULT_CLIENT_OPTIONS.merge(
```
Comment (grosser) [inline lib/samson/secrets/vault_server.rb:76]: FYI could also do:
```suggestion
        VaultClientWrapper.new(
          **DEFAULT_CLIENT_OPTIONS,
```

File: plugins/datadog/test/models/datadog_deploy_event_test.rb [hunk 8]
```diff
@@ -27,13 +27,13 @@ def expected_body(overrides = {})
 
     it 'delivers correct event with deploy updated_at' do
       assert_request(:post, url, with: {body: expected_body}) do
-        deliver(tags: [], time: happened)
+        deliver(**{tags: [], time: happened})
```
Comment (grosser) [inline plugins/datadog/test/models/datadog_deploy_event_test.rb:30]: same sus as the one mentioned before

**Review by grosser — COMMENTED** [review]
looks mostly good, a few `**{foo: "bar"}` calls that look like some automated refactoring mishap

File: .ruby-version (hunk 1 above)
Comment (zendesk-mattlefevre) [reply .ruby-version:1]: We added a bunch of mission versions last week; good call-out, though.

File: app/controllers/concerns/json_renderer.rb (hunk 3 above)
Comment (zendesk-mattlefevre) [reply app/controllers/concerns/json_renderer.rb:60]: Good call-out. I was troubleshooting the new way variables were passed in a few places before realizing the issue was actually with ActiveSupport. I'll clean all that up.

File: app/views/audits/index.html.erb [hunk 9]
```diff
@@ -7,7 +7,7 @@
 
 <%= search_form do %>
   <%= search_select :action, ["create", "update", "destroy"] %>
-  <%= search_select :auditable_type, audited_classes.sort, label: "Resource" %>
+  <%= search_select :auditable_type, audited_classes.sort, **{ label: "Resource" } %>
```
Comment (grosser) [inline app/views/audits/index.html.erb:10]:
```suggestion
  <%= search_select :auditable_type, audited_classes.sort, label: "Resource" %>
```

File: lib/samson/secrets/vault_server.rb [hunk 10]
```diff
@@ -73,13 +73,12 @@ def sync!(other)
 
       def create_client
         VaultClientWrapper.new(
-          DEFAULT_CLIENT_OPTIONS.merge(
+          **DEFAULT_CLIENT_OPTIONS,
             address: address,
             ssl_cert_store: cert_store,
             ssl_verify: tls_verify,
             token: token,
             versioned_kv: versioned_kv?
```
Comment (grosser) [inline lib/samson/secrets/vault_server.rb:81]: these should be on the same level
```suggestion
          address: address,
          ssl_cert_store: cert_store,
          ssl_verify: tls_verify,
          token: token,
          versioned_kv: versioned_kv?
```

File: plugins/datadog/test/models/datadog_monitor_test.rb [hunk 11]
```diff
@@ -25,7 +27,7 @@ def assert_datadog_timeout(&block)
     {
       name: "Monitor Slow foo",
       query: "max(last_30m):max:foo.metric.time.max{*} > 20000",
-      overall_state: "Ok",
+     overall_state: "Ok",
```
Comment (grosser) [inline plugins/datadog/test/models/datadog_monitor_test.rb:30]:
```suggestion
      overall_state: "Ok",
```

File: plugins/datadog/test/models/datadog_monitor_test.rb [hunk 12]
```diff
@@ -41,44 +43,44 @@ def assert_datadog_timeout(&block)
     before { monitor.query = DatadogMonitorQuery.new(match_target: "pod", match_source: "deploy_group.permalink") }
 
     it "returns simple state when asking for global state" do
-      assert_datadog(overall_state: "OK") do
+      assert_datadog(**{overall_state: "OK"}) do
```
Comment (grosser) [inline plugins/datadog/test/models/datadog_monitor_test.rb:46]: these all should get their `**{` removed

File: test/lib/samson/hooks_test.rb [hunk 13]
```diff
@@ -98,7 +98,7 @@ def hooks
 
       Samson::Hooks.only_callbacks_for_plugin('samson_airbrake', :error) do
         hooks.size.must_equal 1
-        Samson::Hooks.fire(:error, mock_exception, foo: 'bar').size.must_equal 1
+        Samson::Hooks.fire(:error, mock_exception, **{foo: 'bar'}).size.must_equal 1
```
Comment (grosser) [inline test/lib/samson/hooks_test.rb:101]:
```suggestion
        Samson::Hooks.fire(:error, mock_exception, foo: 'bar').size.must_equal 1
```

**Review by rsb — APPROVED** [review]

**Comment by grosser** [comment]
unable to reproduce the cleanliness error (was unable to get 3.0 installed so I used 3.1, so maybe if that works bump to that ...)

**Comment by grosser** [comment]
oh ... it's a random error, ran it a few times and it fails :D

**Comment by grosser** [comment]
use this in test/integration/cleanliness_test.rb:8
```
- def all_models
+ let_all(:all_models) do
```

**Pull request merged**